### ISSUE TITLE
Soft particles renderpaths. Is it a good design decision?

### ISSUE THREAD

**Issue by t-artikov** [urho3d/Urho3D#1670]
I have believed that renderpaths are intended for more "fundamental" things like forward/deferred rendering but not for implementing of particular graphical effects. If you implement an another feature in the same way (snow rendering, for example), will you have ForwardSnow.xml, ForwardSPSnow.xml, DeferredSPSnow.xml, etc? 
It will turn the renderpath into a god object.

I don't see reason why soft particles can't be implemented as a technique and work with existing renderpaths (ForwardDepth, Deferred). This renderpaths provide the depth texture and the soft particles shader can read it (like the water shader reads the color texture now).

**Comment by 1vanK** [comment]
You can use renderpath->Append() for your snow effect (like appending postprocesses)

**Comment by 1vanK** [comment]
> This renderpaths provide the depth texture and the soft particles shader can read it (like the water shader reads the color texture now).
Renderpaths shoud send textures to shader like water shader:
```
    <command type="scenepass" pass="refract">
        <texture unit="environment" name="viewport" />
    </command>
```

**Comment by cadaver** [comment]
Reading the depth or not is a pretty fundamental difference. I agree that it's not ideal to have an "explosion" of renderpaths just due to adding a single effect. In practice you would / could also add other effects that depend on depth.

Also remember is that all assets provided by Urho by default are examples or starting points, and quite simple ones at that. In a serious project you likely end up customizing them anyway. You would also be likely to choose one method of rendering (forward, deferred etc.) and stick with it, in which case you could also simplify the techniques you use, because they no longer have to refer to the rendering methods you don't use.

**Comment by cadaver** [comment]
It's true however that we could in fact delete the SP variations of renderpaths and make techniques instead, the reason to do it this way was to allow easy comparing of particles with and without softness. We'd need to bind the depth buffer in the alpha pass when available but I don't think that hurts.

**Comment by 1vanK** [comment]
May be just make separate pass for SP?

**Comment by cadaver** [comment]
It's also possible, but it would hurt back-to-front sorting, since passes are drawn one after one without knowledge of each other, and it's better all alphas are sorted into one pass.

**Comment by t-artikov** [comment]
> Also remember is that all assets provided by Urho by default are examples or starting points, and quite simple ones at that. In a serious project you likely end up customizing your own renderpath.

I believe it will be better if the engine can be customized by adding new code instead of modifying existing one. Otherwise it will be hard for an user to update to a new version of the engine.

Some drawbacks of the renderpaths SP implementation:
* Will SOFTPARTICLES defines be passed to all shaders in the alpha pass, but not only to the particles shader? (this leads to unwanted shader defines combinations)
* There is no fine control when SP are enabled.
  While with a technique we can on/off it per material or based on the Renderer.materialQuailty or based on the LOD.

**Comment by cadaver** [comment]
Not all renderpath usecases can be solved by just adding commands. Postprocesses appended to the end are practically the simplest case. On the other hand, it's unlikely that renderpaths change dramatically as the engine gets updated.

Yes, it is somewhat unnecessary that the SOFTPARTICLES define gets appended to all alpha shaders, however that also means that a variation *without* the define never gets compiled, unless used outside the renderpath (in UI rendering for example.)

However you can count it likely that the SP paths will be dropped in favor of techniques.

**Comment by t-artikov** [comment]
I appreciate that you have accepted my suggestion. Thank you!

**Comment by 1vanK** [comment]
How affects a sending additional texture to shaders on performance? Not every alpha material required depth texture. May be just create separate pass for soft particles?

**Comment by eugeneko** [comment]
_Sampling_ of texture will have performance impact. But just passing.. Not
heavier than each of 100500 texture setups during scene/shadow/etc
rendering, at least at GAPI side.

2016-11-01 0:49 GMT+03:00 1vanK <notifications@github.com>:

**Comment by cadaver** [comment]
Unless you're actually running out of GPU memory and thrashing textures in and out, having an extra texture bound shouldn't have impact. The depth texture is also bound only once in the beginning of the pass, so it doesn't increase the CPU load of drawcall setup.

**Comment by 1vanK** [comment]
SmokeSoft:

![2](https://cloud.githubusercontent.com/assets/13021826/19979125/5cea8e9a-a20a-11e6-93ec-2274a17f4707.png)

With Techniques/DiffUnlitParticleAlphaSoftExpand.xml plane is full invisible

**Comment by 1vanK** [comment]
Should DiffUnlitParticleAlphaSoft enable alphatest ?

![2](https://cloud.githubusercontent.com/assets/13021826/19979546/eba09700-a20b-11e6-9ffa-0ebf3ae0efbd.png)

![2 -](https://cloud.githubusercontent.com/assets/13021826/19979550/ee265fbe-a20b-11e6-8b2b-ab5c81a222c9.png)

**Comment by 1vanK** [comment]
With enabled depthtest

![2](https://cloud.githubusercontent.com/assets/13021826/19979619/291e956e-a20c-11e6-86f7-3f5d372bad04.png)
```
<technique vs="UnlitParticle" ps="UnlitParticle" vsdefines="VERTEXCOLOR SOFTPARTICLES" psdefines="DIFFMAP VERTEXCOLOR SOFTPARTICLES">
    <pass name="alpha" depthwrite="false" depthtest="true" blend="alpha" />
</technique>
```

**Comment by cadaver** [comment]
Yes. This was a copying mistake when the techniques were adapted from the "expand" variations. "Shrinking" soft particle techniques all should have depth test on.

**Comment by 1vanK** [comment]
Also with enabled depth test DiffZ always >=0, so in
```
float diffZ = abs(depth - particleDepth) * (cFarClipPS - cNearClipPS);
```
"abs" is unnecessary
